### PR TITLE
fixes TextInputFormatter gets wrong old value of a selection

### LINKED ISSUE #75502: TextInputFormatter gets wrong old value of a selection
TextInputFormatter gets wrong old value of a selection after character deletion when keyboardType is TextInputType.phone.

## Example

```
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() => runApp(App());

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: SafeArea(
          child: TextField(
            controller: TextEditingController()..text = '123',
            inputFormatters: [SomeFormatter()],
            keyboardType: TextInputType.phone,
          ),
        ),
      ),
    );
  }
}

class SomeFormatter extends TextInputFormatter {
  @override
  TextEditingValue formatEditUpdate(TextEditingValue oldValue, TextEditingValue newValue) {
    print(oldValue);
    return newValue;
  }
}
```

**Steps to reproduce:** run example, tap on text field, press backspace ones

**Expected result:** TextEditingValue(text: ┤123├, selection: TextSelection(baseOffset: **3**, extentOffset: **3**, affinity: TextAffinity.upstream, isDirectional: false), composing: TextRange(start: -1, end: -1))

**Actual result:** TextEditingValue(text: ┤123├, selection: TextSelection(baseOffset: **2**, extentOffset: **2**, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))

Tested on version 1.26, probably reproduced since version 1.23.

<details>
  <summary>Flutter Doctor</summary>
  
```
[√] Flutter (Channel dev, 1.26.0-17.2.pre, on Microsoft Windows [Version 10.0.19041.746], locale ru-RU)
[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[√] Chrome - develop for the web
[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.5.2)
[√] Android Studio (version 4.1.0)
```
</details>

### PR DESCRIPTION
**Pull request by chunhtai**
Currently EditableText listens to text editing value changes from its RenderEditable in two ways:

1. The renderObject.onSelectionChange callback.
2. The EditableText.textEditingValue setter.

I believe that the original intent of these two methods was:

1. The `onSelectionChange` method is only called when a new selection has been created; the EditableText rebuilds the selection overlay when it is called.
2. `EditableText.textEditingValue` is called when the selection changes from an old value to a new value, and the EditableText simply updates the selection value in overlay without rebuilding it.

RenderEditable becomes confused when the value changes, sometimes, both methods may need to be called... This causes another issue because they will mess up each other.

If renderObject.onSelectionChange is called before EditableText.textEditingValue, the EditableText.textEditingValue will not work properly. The  EditableText.textEditingValue will need to notify the inputformatter about value changes, it will send the value to the inputformatter before and after the update. However, the selection has already been updated by the enderObject.onSelectionChange at this point, it does not know what's the selection in the old value. This is the cause of this issue https://github.com/flutter/flutter/issues/75502.

if the EditableText.textEditingValue is called before renderObject.onSelectionChange, the renderObject.onSelectionChange will be ignored and will cause the selection overlay to not be built.

Combining all of the above reasons, I created this PR to make editableText only listen to EditableText.textEditingValue (I changes it to updateTextEditingValue because it can't no longer be just a setter), and make renderEditable always call updateTextEditingValue when the selection changes. On top of that. I want to deprecate the renderObject.onSelectionChange because it is not really useful to editableText.

There is currently no test, I will add more test once we are sure this is the path we want to proceed.

fixes https://github.com/flutter/flutter/issues/75502.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 17 untriaged digest(s) on patchset 1.
View them at https://flutter-gold.skia.org/cl/github/75541

**Comment by skia-gold** [comment]
Gold has detected about 3 untriaged digest(s) on patchset 2.
View them at https://flutter-gold.skia.org/cl/github/75541

**Comment by skia-gold** [comment]
Gold has detected about 3 untriaged digest(s) on patchset 3.
View them at https://flutter-gold.skia.org/cl/github/75541

**Comment by skia-gold** [comment]
Gold has detected about 6 untriaged digest(s) on patchset 4.
View them at https://flutter-gold.skia.org/cl/github/75541

**Review by justinmc — COMMENTED** [review]
I think I'm on board with this approach assuming tests pass and everything works the same as before. The setter and onSelectionChanged do seem to be redundant and their differences are confusing.

Do we know if anyone besides Flutter itself has been using onSelectionChanged?  It seems like there will no longer be a straightforward way to listen for selection changes.  Users would have to use updateTextEditingValue and then check if the new value is different, or something like that.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -33,37 +33,6 @@ const Radius _kFloatingCaretRadius = Radius.circular(1.0);
 /// Used by [RenderEditable.onSelectionChanged].
 typedef SelectionChangedHandler = void Function(TextSelection selection, RenderEditable renderObject, SelectionChangedCause cause);
 
-/// Indicates what triggered the change in selected text (including changes to
-/// the cursor location).
-enum SelectionChangedCause {
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/editable.dart:38]: I think that will be a layer violation, text_input.dart is in service layer

**Comment by chunhtai** [comment]
@justinmc unfortunately I know there are a lot of library out there just copied the entire EditableText, (sometimes also RenderEditable) and makes changes to fit their need, so the changes to the textEditingValue setter will be a big breaking change for them. I think this is something that needed to be done anyway, and i am glad you thought the same. I am curious on the text refactoring project, will it also be a big breaking change too? If so, should we coordinate the merge so we only break them once?

Thanks for the feedback, I will update the PR and add more tests

**Comment by LongCatIsLooong** [comment]
@justinmc Is this still needed once we move key events handling out? I'm under the impression that once we're done `RenderEditable` won't be able to make changes to its `TextEditingValue`?

**Comment by skia-gold** [comment]
Gold has detected about 6 untriaged digest(s) on patchset 10.
View them at https://flutter-gold.skia.org/cl/github/75541

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 2]
```diff
@@ -102,6 +102,7 @@ class _CupertinoTextFieldSelectionGestureDetectorBuilder extends TextSelectionGe
 
   @override
   void onSingleTapUp(TapUpDetails details) {
+    editableText.hideToolbar();
```
Comment (chunhtai) [inline packages/flutter/lib/src/cupertino/text_field.dart:105]: We have this in Material Textfield but not cupertinotextfield, This is to deal with the case where there are toolbar open and user tap the text outside of the current selection. The expect behavior is the tool bar will be closed and the collapsed selection moved to a new place with the toolbar remaining closed.

The workflow is as the following:

onSingleTapUp(hide toolbar) -> EditableTextOnselectionChange(update toolbar) -> composing changes(hide toolbar)

In theory, if we don't add this hide toolbar, you should see the selection overlay gets moved to the new location with the toolbar open. the toolbar will then be hidden after a frame delay.


This works prior to this pr in a weird way. When user tap and the toolbar will be rebuild with the new selection, but that will only show up after one frame delay. right before the frame end, the selection change again probably due to composing changes. This will hide the toolbar before it builds.

In short, we have a selection overlay built and destroyed in the same frame.

My PR optimizes the selection change that it won't need to rebuild the selection overlay every time. This breaks because the overlay will get updated immediately before the hide toolbar remove the overlay on next frame. Visually, you will see the selection change to the new place with the toolbar still open, and the toolbar will close after one frame.

Comment (chunhtai) [reply packages/flutter/lib/src/cupertino/text_field.dart:105]: This is also part of the reason why i think we should move the selection overlay out of editable text.

File: packages/flutter/test/cupertino/text_field_test.dart [hunk 3]
```diff
@@ -3475,7 +3475,6 @@ void main() {
         from: tester.getTopRight(find.byType(CupertinoApp)),
         cause: SelectionChangedCause.tap,
       );
-      expect(state.showToolbar(), true);
```
Comment (chunhtai) [inline packages/flutter/test/cupertino/text_field_test.dart:3478]: Prior to this pr the entire selection overlay will be rebuilt when the editableText rebuild with different selection.

I feel that is weird. if you have a selection with toolbar and handles on a textfield and you update the texteditingcontroller with different selection, you will see the selection gets moved to a new place with the handles but no toolbar.

After this PR, the toolbar will remain open when the selection moved to the new place.

That means we don't need to call the showtoolbar here in the test.

File: packages/flutter/test/material/text_field_focus_test.dart [hunk 4]
```diff
@@ -119,6 +119,8 @@ void main() {
     expect(tester.testTextInput.isVisible, isTrue);
 
     tester.testTextInput.hide();
+    final EditableTextState state = tester.state<EditableTextState>(find.byType(EditableText));
+    state.connectionClosed();
```
Comment (chunhtai) [inline packages/flutter/test/material/text_field_focus_test.dart:123]: This test is testing a non realistic situation.

In real app, closing the input connection will cause the textfield to be unfocus, tapping it again, the textfield will regain focus and open the input connection.

This works before because it relies on the bug that the selection does not update correctly in rendereditable. So everytime you tap, you receive a new selection which cause a selection changes and open the keyboard.

File: packages/flutter/test/widgets/editable_text_cursor_test.dart [hunk 5]
```diff
@@ -49,6 +49,7 @@ void main() {
   });
 
   testWidgets('cursor layout has correct width', (WidgetTester tester) async {
+    EditableText.debugDeterministicCursor = true;
```
Comment (chunhtai) [inline packages/flutter/test/widgets/editable_text_cursor_test.dart:52]: The pr shifted the blinking timer a bit, i figured the blinking is not necessary for this test

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 6]
```diff
@@ -5357,6 +5357,7 @@ void main() {
       'TextInput.setEditingState',
       'TextInput.setEditingState',
       'TextInput.show',
+      'TextInput.show',
```
Comment (chunhtai) [inline packages/flutter/test/widgets/editable_text_test.dart:5360]: and additional show is called due to selection change which was not called before, this should be a bug.

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 7]
```diff
@@ -5471,16 +5472,18 @@ void main() {
     log.clear();
 
     final EditableTextState state = tester.firstState(find.byType(EditableText));
-
     // setEditingState is not called when only the remote changes
-    state.updateEditingValue(const TextEditingValue(
+    state.updateEditingValue(TextEditingValue(
       text: 'a',
+      selection: controller.selection,
```
Comment (chunhtai) [inline packages/flutter/test/widgets/editable_text_test.dart:5478]: if the selection is not set, it will be set to collapsed at -1, which should cause a selection change. This is not needed before because it previously just ignored the selection change

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 8]
```diff
@@ -6619,6 +6629,7 @@ void main() {
     final EditableTextState state = tester.state<EditableTextState>(find.byType(EditableText));
     state.updateEditingValue(const TextEditingValue(
       text: 'foo composing bar',
+      selection: TextSelection.collapsed(offset: 4),
```
Comment (chunhtai) [inline packages/flutter/test/widgets/editable_text_test.dart:6632]: selection needed to be within composing range, otherwise the composing will be reseted by texteditingcontroller. It is not needed before because the it does not fire selection changes prior to this pr

**Comment by chunhtai** [comment]
Hi @justinmc this pr is ready, can you take another look? thanks!

**Comment by chunhtai** [comment]
a friendly bump

**Review by LongCatIsLooong — COMMENTED** [review]
Left a few questions.

An alternative to this is to use `EditableText`'s `batchEdit` function, storing the old value in `beginBatchEdit` and call the user callbacks in `endBatchEdit`. But the call paths look really convoluted and thanks for cleaning it up!

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 9]
```diff
@@ -2271,6 +2274,16 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     // sending multiple `TextInput.updateEditingValue` messages.
     beginBatchEdit();
     _value = value;
+    // Changes made by the keyboard can sometimes be "out of band" for listening
+    // components, so always send those events, even if we didn't think it
+    // changed. Also, focusing an empty field is sent as a selection change even
+    // if the selection offset didn't change.
+    if (selectionChanged ||
+        (userInteraction &&
+        ((value.selection.baseOffset == 0 && value.selection.extentOffset == 0 && !_hasFocus) ||
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/editable_text.dart:2283]: this is for corner case where the textfield is active and user single tap the textfield again to show the selection overlay.
Therefore, we only want to build overlay when it is trigger by the user if the selection is still at 0

File: packages/flutter/lib/src/services/text_input.dart [hunk 10]
```diff
@@ -768,10 +810,10 @@ abstract class TextSelectionDelegate {
...
   /// See also:
   ///
-  /// * [EditableTextState.textEditingValue]: an implementation that applies
-  ///   additional pre-processing to the specified [value], before updating the
-  ///   text editing state.
-  set textEditingValue(TextEditingValue value);
+  /// * [EditableTextState.userUpdateTextEditingValue]: an implementation that
+  ///   applies additional pre-processing to the specified [value], before
+  ///   updating the text editing state.
+  void userUpdateTextEditingValue(TextEditingValue value, SelectionChangedCause? cause);
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/services/text_input.dart:816]: probably should be not nullable, thanks for catching

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/editable_text.dart:2283]: ah yea i should check for selection change before calling onselectionchange

**Review by justinmc — APPROVED** [review]
LGTM

Sorry I missed this notification!  Thanks for doing this clean up.  I'm interested in the batch idea by @LongCatIsLooong above but it should be fine either way.

My text editing refactoring is trying to be minimally breaking, so I think we shouldn't need to coordinate.  The first PR is https://github.com/flutter/flutter/pull/75032 and it doesn't contain any breaking changes.  As I port some of the tap events that affect the stuff in this PR then I might want to break some of the API there, but not yet.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:2283]: why the `!_hasFocus` if the context menu should only be updated on a focused field?

[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/editable_text.dart:2283]: ok this is a series of legacy code and bug make me think this is correct. We want to build the selection overlay when user type keyboard or long press. and we need to want to call the onselectionchange because textfield relies on it to decide whether it really want to show selection handles. (I know it is weird, but again, this why i think we should move selectionoverlay outside..). It has already like this prior to the change, that longpress on a textfield will fire selection change even if selection does not change.

**Comment by chunhtai** [comment]
Thanks for the review, I will proceed to write breaking change doc for the deprecated callbacks

**Comment by chunhtai** [comment]
oh it looks like we don't consider breaking change f no customer tests fail. I guess we should be fine then.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/editable_text.dart]: I update it to check for long press instead

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 11]
```diff
@@ -2271,6 +2274,16 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
     _value = value;
+    // Changes made by the keyboard can sometimes be "out of band" for listening
+    // components, so always send those events, even if we didn't think it
+    // changed. Also, the user long pressing should always send a selection change
+    // as well.
+    if (selectionChanged ||
+        (userInteraction &&
+        (cause == SelectionChangedCause.longPress ||
+         cause == SelectionChangedCause.keyboard))) {
+      _handleSelectionChanged(value.selection, cause);
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:2285]: nit: could you add a TODO since ideally this shouldn't call `widget.onSelectionChanged`?

**Review by LongCatIsLooong — APPROVED** [review]

**Comment by sgehrman** [comment]
The stable branch in Flutter 2 has this bug now.  Could we get this fixed?

**Comment by chunhtai** [comment]
> The stable branch in Flutter 2 has this bug now. Could we get this fixed?

@sgehrman This change was reverted right before the stable release, but it is now relanded. I will see if we can cherry pick this.
